---
title: "Priors_production"
author: "karine Durand"
date: "23 octobre 2018"
update: "13 décembre 2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## production des priors SI

``

####mon_script
```{r}
library(stats)
library("KScorrect", lib.loc="~/R/x86_64-pc-linux-gnu-library/3.3")
######partie locus 
#variables locus
#-L=taille du gene
#-t=theta
#-r=rho
#-delta=taille du track recombinant
#boucle de 1000000 iterations(1000000 tirage demographique)
demo<-NULL
locus<-NULL
tbs<-NULL
#####TIRER un prior locus dans une distribution uniforme de bornes 
L<-scan("/home/kadurand/partage_windows/Xylella/analyses_genomiques/ABC/1368oRTHOLOGUES_summarystats/lenght")#bound_taille du gene[100-2500]
t<-runif(1368,0, 0.001 )#bound_theta=[0-0.0003]bornes vrai pour 13pauca_multiplex augmenter la borne sup à 0.001]
r<-runif(1368,0,0.001)##bound_rho=[0,0.0006]bornes vrai pour 13pauca_multiplex augmenter la borne sup à 0.001]
delta<-round(runif(1368,10, 1000))#bound=[10-1000]
#print(L,t,r,delta)
m_locus=matrix(c(L,t,r,delta),ncol=4)
m_locus=as.data.frame(m_locus)

for (i in 1:10000){#tirage des priors demographiques
#variables demographique modéle SI
#Param_demo (3) =   Ts, N1, N2,
Ts<-rlunif(1,100,100000)#bound=[100-100000]
N1<-rlunif(1,100,1000000)#bound=[100-1000000]
N2<-rlunif(1,100,1000000)#bound=[100-1000000]
Na<-rlunif(1,100,1000000)#bound=[100-1000000]
 
#print( N1, N2, Ts)
m_demo=matrix(c(Ts,N1,N2,Na),ncol=4)
m_demo=as.data.frame(m_demo)
locus<-cbind(m_locus,m_demo)

path <- "/home//kadurand/partage_windows/Xylella/analyses_genomiques/ABC/fastSimBac_linux/Priors_SI_10000/priorSI-" 
write.table(locus,file= paste(path,i, sep="-"),col.names=FALSE,row.names =FALSE)
}
```

