---
title: "popgenome.Rmd"
author: "karine Durand"
date: "11 avril 2019"
output: html_document
---



```{r cars}
library(PopGenome)
setwd("~/partage_windows/Xylella/analyses_genomiques/ABC/myscripts/popgenome/")
# Load the data
snp <- readData("~/partage_windows/Xylella/analyses_genomiques/ABC/myscripts/popgenome", format="VCF")
aln<- readData("~/partage_windows/Xylella/ORTHOLOGUES/19pauca_multiplex/1338_orthologues_gblock/", format="fasta",SNP.DATA=FALSE,FAST=T)
# This is complex object, with several slots
get.sum.data(snp)
show.slots(snp)
snp@n.sites
# Set populations
pops <- get.individuals(snp)[[1]]
pops <- get.individuals(aln)[[1]]
paucaSA <- c("xylefa8072","xylefa8429pac","xylefaCOF7","xylefaOLS8","xylefaOLS9","xylefacodi")
paucaCA <- c("xylefa9a5c","xylefa1199","xylefa3124","xylefaco6c","xylefaco32","xylefaCOF4","xylefaCVC1","xylefaCVC6","xylefaHib4","xylefaJa12","xylefaPr8x","xylefasFb7","xylefaU24D")
multi <- c("xylefa3571","xylefa4311","xylefa8070pac","xylefa8416pac","xylefa8417","xylefa8418pac","xylefa8430pac","xylefa8431pac","xylefa8432pac","xylefa8433pac","xylefa8434pac","xylefa8435pac","xylefaBB01","xylefadixn","xylefagrif","xylefaM12s","xylefaSyVa")
pauca<- c("xylefa9a5c","xylefa1199","xylefa3124","xylefaco6c","xylefaco32","xylefaCOF4","xylefaCVC1","xylefaCVC6","xylefaHib4","xylefaJa12","xylefaPr8x","xylefasFb7","xylefaU24D","xylefa8072","xylefa8429pac","xylefaCOF7","xylefaOLS8","xylefaOLS9","xylefacodi")

snp  <- set.populations(snp, list(paucaCA ,paucaSA,multi))
snp@populations
aln  <- set.populations(aln, list(paucaCA ,paucaSA,multi))
aln@populations
```

## Let's calculate FST between the two populations and nucleotide diversity in each of the populations.



```{r pressure}
get.sum.data(aln)
get.neutrality(aln)[[1]]
aln  <- diversity.stats(aln)
aln@nuc.diversity.within
aln@region.stats@nucleotide.diversity
aln <- neutrality.stats(aln) 
win_aln<- neutrality.stats(win_aln)
win_aln@n.sites 
win_aln@Pi
win_aln@Tajima.D
write()
aln@hap.diversity.within->hapdiversity
aln@hap.diversity.between->hapdiversitybetween
aln@haplotype.counts
aln@Pi
aln  <- recomb.stats(aln)
aln@get.recomb
X<-c(10000,20000,30000,40000,50000,60000,70000,80000,90000,100000,110000,120000,130000,140000,150000,160000,170000,180000,190000,200000,210000,220000,230000,240000,250000,260000,270000,280000,290000,300000,310000,320000,330000,340000,350000,360000,370000,380000,390000,400000,410000,420000,430000,440000,450000,460000,470000,480000,490000,500000,510000,520000,530000,540000,550000,560000,570000,580000,590000,600000,610000,620000,630000,640000,650000,660000,670000,680000,690000,700000,710000,720000,730000,740000,750000,760000,770000,780000,790000,800000,810000,820000,830000,840000,850000,860000,870000,880000,890000,900000,910000,920000,930000,940000,950000,960000,970000,980000,990000,1000000,1010000,1020000,1030000,1040000,1050000,1060000,1070000,1080000,1090000,1100000,1110000,1120000,1130000,1140000,1150000,1160000,1170000,1180000,1190000,1200000,1210000,1220000,1230000,1240000,1250000,1260000,1270000,1280000,1290000,1300000,1310000,1320000,1330000,1340000,1350000,1360000,1370000,1380000,1390000,1400000,1410000,1420000,1430000,1440000,1450000,1460000,1470000,1480000,1490000,1500000,1510000,1520000,1530000,1540000,1550000,1560000,1570000,1580000,1590000,1600000,1610000,1620000,1630000,1640000,1650000,1660000,1670000)
as.data.frame(win_aln@Tajima.D)->datataj
plot(X, datataj[,1],xlab="position (Mb)",ylab="nucleotide diversity", main = "Chromosome 2L (10kb windows)")
snp <- F_ST.stats(snp) 
get.F_ST(snp)[[1]]
get.F_ST(snp)[[2]] 
get.F_ST(snp)[[3]] # each line is a scaffold
snp@nucleotide.F_ST
```

 Print diversities
 

```{r}
get.diversity(snp)
get.diversity(snp)[[1]] # paucaCA 
get.diversity(snp)[[2]] # ,paucaSA
get.diversity(snp)[[3]] #,multi
snp@nuc.diversity.within
```
Another useful tool is to do the calculations along a sliding window.

```{r}
win_snp <- sliding.window.transform(snp, 
    width=10000, jump=10000, 
    type=2,
    whole.data=FALSE)

win_aln <- sliding.window.transform(aln, 
    width=10000, jump=10000, 
    type=2,
    whole.data=FALSE)
length(win_aln@region.names)
#statistics
win_aln <- diversity.stats(win_aln)

#
nucdiv <- win_aln@nuc.diversity.within
as.data.frame(nucdiv)->pi
plot(X, nucdiv[,1], xlab="position (Mb)",ylab="nucleotide diversity", main = "Chromosome 2L (10kb windows)")


# Measurements per window
win_aln <- F_ST.stats(win_aln)
pairwise.FST <- t(win_aln@nuc.F_ST.pairwise)

plot(pairwise.FST, type = "l", xaxt="n", xlab="position (Mb)",ylab="pairwiseFST", main = "Chromosome 2L (10kb windows)")
x=10000
for i 
ids<-
win_snp@nucleotide.F_ST
win_aln@nuc.diversity.within

# A simple plot
win_fst <- win_snp@nucleotide.F_ST[,1]
paucaCA_div  <- win_snp@nuc.diversity.within[,1] # diversity among B (bb = "big B")
paucaSA_div  <- win_snp@nuc.diversity.within[,2] # diversity among B (lb = "little b")
multi_div  <- win_aln@nuc.diversity.within[,2]
```



```{r}
plot(1:length(win_aln), win_aln)

par(mfrow=c(2,1))
win_fst <- win_aln@nucleotide.F_ST[,1]
plot(1:1671846, win_aln@nuc.diversity.within[,1])

win_fst <- win_snp@nucleotide.F_ST[,1]
plot(1:length(multi_div ), multi_div )
```

# calcul R2 
```{r}

snp <- calc.R2(snp )# show the result:# GENOME.class@region.stats@linkage.disequilibrium# [[x]][[y]] x:region, y:population
```
